base/rational.jl

```diff
@@ -129,12 +129,344 @@ Bool(x::Rational) = x==0 ? false : x==1 ? true :
+# even.
+#
+# `requested_precision` must be positive.
+function rational_to_float_impl(
+    x::Rational{S},
+    ::Type{T},
+    requested_precision::Integer,
+    to_float::C,
+) where {C<:Union{Type,Function},S<:Integer,T<:Integer}
+    s = Int8(sign(x))
+    a = abs(x)
+
+    num = numerator(a)
+    den = denominator(a)
+
+    # Handle special cases
+    num_is_zero = iszero(num)
+    den_is_zero = iszero(den)
+    if den_is_zero
+        num_is_zero && return to_float(NaN)
+        return to_float(s * Inf)
+    end
+    num_is_zero && return to_float(false)
+
+    components = rational_to_float_components(
+      num,
+      den,
+      requested_precision,
+      T,
+      RationalToFloatRoundingMode.round_to_nearest_ties_to_even,
+    )
+    mantissa = to_float(components.mantissa)
+
+    # TODO: `ldexp` could be replaced with a mere bit of bit twiddling
+    # in the case of `Float16`, `Float32`, `Float64`
+    ret = ldexp(s * mantissa, components.exponent)
+
+    # TODO: faster?
+    if iszero(ret) | issubnormal(ret)
+        # "Rounding to odd" to prevent double rounding error, see
+        # https://hal-ens-lyon.archives-ouvertes.fr/inria-00080427v2
+        components = rational_to_float_components(
+          num,
+          den,
+          requested_precision,
+          T,
+          RationalToFloatRoundingMode.round_to_zero,
+        )
+        mantissa = to_float(components.mantissa | !components.is_exact)
+
+        # TODO: `ldexp` could be replaced with a mere bit of bit
+        # twiddling in the case of `Float16`, `Float32`, `Float64`
+        ret = ldexp(s * mantissa, components.exponent)
+    end
+
+    ret
+end
+
+rational_to_float_promote_type(
+    ::Type{F},
+    ::Type{S},
+) where {F<:AbstractFloat,S<:Integer} =
+    BigInt
+
+rational_to_float_promote_type(
+    ::Type{F},
+    ::Type{S},
+) where {F<:AbstractFloat,S<:Unsigned} =
+    rational_to_float_promote_type(F, signed(S))
+
+# As an optimization, use a narrower type when possible.
+rational_to_float_promote_type(
+    ::Type{Float16},
+    ::Type{<:Union{Int8,Int16}},
+) =
+    Int32
+rational_to_float_promote_type(
+    ::Type{Float32},
+    ::Type{<:Union{Int8,Int16}},
+) =
+    Int64
+rational_to_float_promote_type(
+    ::Type{Float64},
+    ::Type{<:Union{Int8,Int16}},
+) =
+    Int128
+rational_to_float_promote_type(
+    ::Type{<:Union{Float16,Float32}},
+    ::Type{Int32},
+) =
+    Int64
+rational_to_float_promote_type(
+    ::Type{Float64},
+    ::Type{Int32},
+) =
+    Int128
+rational_to_float_promote_type(
+    ::Type{<:Union{Float16,Float32,Float64}},
+    ::Type{Int64},
+) =
+    Int128
+
+rational_to_float(
+    x::Rational{S},
+    ::Type{F},
+) where {F<:AbstractFloat,S<:Integer} = rational_to_float_impl(
+    x,
+    rational_to_float_promote_type(F, S),
+    precision(F),
+    F,
+)
+
+(::Type{F})(x::Rational{S}) where {F<:AbstractFloat,S<:Integer} =
+    rational_to_float(x, F)
```

Comment:
Why the additional forward to an `_impl` function? Both of these are internal either way.